File: packages/python/qsu/array/sortByObjectKey.py
from .sortNumeric import _naturalKey


def sortByObjectKey(
	array: list,
	key: str,
	descending: bool = False,
	numerically: bool = False,
) -> list:
	# Sort a copy: `list.sort` reorders in place. Use `reverse=` rather than reversing the
	# result, which would also flip the order of equal elements.
	if numerically:
		return sorted(array, key=lambda item: _naturalKey(item[key]), reverse=descending)

	return sorted(array, key=lambda item: item[key], reverse=descending)

File: packages/javascript/lib/array/sortByObjectKey.ts
import { compareNaturalKey, naturalKey } from './_naturalCompare.js';

export function sortByObjectKey(
	array: any[],
	key: string,
	descending = false,
	numerically = false
): any[] {
	// Sort a copy: `Array.prototype.sort` reorders in place. Flip the comparison for
	// descending order instead of reversing, which would also flip equal elements.
	if (numerically) {
		// Build each key once rather than once per comparison.
		const decorated = array.map((item) => ({ item, sortKey: naturalKey(item[key]) }));

		decorated.sort((a, b) => {
			const order = compareNaturalKey(a.sortKey, b.sortKey);

			return descending ? -order : order;
		});

		return decorated.map((entry) => entry.item);
	}

	return [...array].sort((a: any, b: any) => {
		if (!descending) {
			if (a[key] < b[key]) {
				return -1;
			}
			if (a[key] > b[key]) {
				return 1;
			}

			return 0;
		}

		if (a[key] > b[key]) {
			return -1;
		}
		if (a[key] < b[key]) {
			return 1;
		}

		return 0;
	});
}
